KEYCLOAK-16755 ClearExpiredUserSessions optimization. Rely on infinis…

mposolda opened the pull request:
…pan expiration rather than Keycloak own background task.

<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

mposolda committed:
KEYCLOAK-16755 ClearExpiredUserSessions optimization. Rely on infinispan expiration rather than Keycloak own background task.

hmlnarik commented on server-spi/src/main/java/org/keycloak/models/UserSessionProvider.java:
```diff
@@ -138,6 +138,14 @@ UserSessionModel createUserSession(String id, RealmModel realm, UserModel user,
+     * Remove expired events in all the realms.
+     * Implementation doesn't need to propagate removal of expired userSessions to userSessionPersister. Cleanup on persister will be called separately
+     */
+    default void removeAllExpired(KeycloakSession session) {
```
Is it necessary to introduce `KeycloakSession` parameter? `UserSessionProvider` instance is always aware of the session it was created with. Seems like this would introduce an exception to the rest of the methods.

hmlnarik reviewed (commented):
Looks like a great addition, thank you! I've not yet done full review, but found one small thing, please see inline.

mposolda commented:
@hmlnarik Thanks for the comment. I've updated the PR per your suggestion. I am planning few more minor updates to the PR (Some cleanup and reduce some info logging). I've did also try pipeline run with the base, cluster and cross-dc tests https://keycloak-jenkins.com/view/pipelines/job/universal-test-pipeline-server/2530/ . There are no regressions in comparison to master.

Will do some more changes and will probably try full pipeline run. Will coment once it is ready for the review.

mposolda commented:
@hmlnarik @martin-kanis Few more notes to what we discussed yesterday:

- I changed javadoc of UserSessionPRovider.removeAllExpired and removed the note about persister as we discussed yesterday. I did not change javadoc for `UserSessionProvider.removeExpired(RealmModel realm)` . I am leaving that to @martin-kanis to do that in the PR, which moves the persister call "inside" UserSessionProvider. 

- I've tried some performance testing on my laptop for online/offline sessions and clientSessions with and without the cross-dc environment. The infinispan based expiration works as expected and entities are correctly cleared from the cache. And also from the remoteCache in case of the cross-dc environment. Entries are added to remoteCache with expiration and maxIdle, which in some cases is longer than the expiration/maxIdle of entries added to the distributed cache. This is due to UserSessionModel.setLastSessionRefresh and clientSession.setTimestamp, where we want to avoid updating remoteCache for each refresh token calls. However expiration of entries from the distributed cache also propagates expiration of such entries to the remoteCache due the remoteStore.

mposolda commented:
@hmlnarik @martin-kanis This should be ready for the review. Do you please have a chance to look at it? I am still waiting for the pipeline run to finish, but hopefully there won't be any regression..

martin-kanis commented on server-spi/src/main/java/org/keycloak/sessions/AuthenticationSessionProvider.java:
```diff
@@ -71,6 +72,11 @@ default RootAuthenticationSessionModel createRootAuthenticationSession(String id
     void removeRootAuthenticationSession(RealmModel realm, RootAuthenticationSessionModel authenticationSession);
 
+    /**
+     * Remove expired events in all the realms
```
Should be here rather something like "Remove expired authentication sessions in all the realms" ?

mposolda replied:
Done

martin-kanis commented on server-spi/src/main/java/org/keycloak/models/UserSessionProvider.java:
```diff
@@ -138,6 +138,11 @@ UserSessionModel createUserSession(String id, RealmModel realm, UserModel user,
     void removeUserSessions(RealmModel realm, UserModel user);
 
+    /**
+     * Remove expired events in all the realms.
```
Should be here rather something like "Remove expired user sessions in all the realms" ?

mposolda replied:
Done. I've added "Remove expired user sessions and client sessions in all the realms" as it applies to client sessions as well.

martin-kanis commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/SessionTimeouts.java:
```diff
@@ -0,0 +1,293 @@
+            return ENTRY_EXPIRED_FLAG;
+        }
+
+        return timeToExpire * 1000;
```
I would extract the multiplication to a method called "toMillis()" or something similar.

mposolda replied:
I figured we already have the generic util method "Time.toMillis" for this, so I've just used that one :) Just changed the argument type to "long" for Time.toMillis (it was int before).

martin-kanis commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java:
```diff
@@ -245,47 +247,48 @@ protected void checkRemoteCaches(KeycloakSession session) {
-        boolean sessionsRemoteCache = checkRemoteCache(session, sessionsCache, (RealmModel realm) -> {
+        RemoteCache sessionsRemoteCache = checkRemoteCache(session, sessionsCache, (RealmModel realm) -> {
             // We won't write to the remoteCache during token refresh, so the timeout needs to be longer.
             return realm.getSsoSessionMaxLifespan() * 1000;
```
Similar to note below. Consider an extraction of multiplication to a separate method in this class.

mposolda replied:
Done with "Time.toMillis"

martin-kanis reviewed (changes requested):
Thank you Marek. Overall it looks good to me. I just added a few minor notes, see inline.

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LoginTest.java:
```diff
@@ -770,23 +770,29 @@ public void loginExpiredCode() {
 
-        Assert.assertEquals("Your login attempt timed out. Login will start from the beginning.", loginPage.getError());
-        setTimeOffset(0);
+            //loginPage.assertCurrent();
```
Done

martin-kanis replied:
I meant this commented line :)
```suggestion
```

mposolda commented:
@martin-kanis Thanks for the review! I've addressed your suggestions. Details in the comments.

hmlnarik commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/SessionTimeouts.java:
```diff
@@ -0,0 +1,293 @@
+     * @param userSessionEntity
+     * @return
+     */
+    public static Long getUserSessionLifespanMs(RealmModel realm, UserSessionEntity userSessionEntity) {
```
Should this return `long` rather than `Long`? It seems never to return `null`. (similarly for the rest of this class)

hmlnarik commented on testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/model/infinispan/InfinispanTestUtil.java:
```diff
@@ -0,0 +1,94 @@
+    private static TimeService origTimeService = null;
+
+    /**
+     * Set Keycloak test TimeService to infinispan cacheManager. This will cause that infinispan will be aware of Keycloak Time offset, which is useful
+     * for testing that infinispan entries are expired after moving Keycloak time forward with {@link org.keycloak.common.util.Time#setOffset} .
```
I like this very much!

hmlnarik commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanKeycloakTransaction.java:
```diff
@@ -155,12 +155,12 @@ public String toString() {
                 @Override
                 public String toString() {
-                    return String.format("CacheTaskWithValue: Operation 'replace' for key %s", key);
+                    return String.format("CacheTaskWithValue: Operation 'replace' for key %s, lifespan %d TimeUnit %s", key, lifespan, lifespanUnit.toString());
```
explicit `toString()` is not necessary, `%s` would do this automatically
```suggestion
                    return String.format("CacheTaskWithValue: Operation 'replace' for key %s, lifespan %d TimeUnit %s", key, lifespan, lifespanUnit);
```

hmlnarik commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java:
```diff
@@ -245,47 +248,48 @@ protected void checkRemoteCaches(KeycloakSession session) {
-        checkRemoteCache(session, offlineClientSessionsCache, (RealmModel realm) -> {
-            return realm.getOfflineSessionIdleTimeout() * 1000;
-        });
+        RemoteCache offlineClientSessionsRemoteCache = checkRemoteCache(session, offlineClientSessionsCache, (RealmModel realm) -> {
```
Is this variable used?

hmlnarik commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java:
```diff
@@ -245,47 +248,48 @@ protected void checkRemoteCaches(KeycloakSession session) {
 
         Cache<UUID, SessionEntityWrapper<AuthenticatedClientSessionEntity>> clientSessionsCache = ispn.getCache(InfinispanConnectionProvider.CLIENT_SESSION_CACHE_NAME);
-        checkRemoteCache(session, clientSessionsCache, (RealmModel realm) -> {
+        RemoteCache clientSessionsRemoteCache = checkRemoteCache(session, clientSessionsCache, (RealmModel realm) -> {
```
Is this variable used?

hmlnarik commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remotestore/RemoteCacheSessionListener.java:
```diff
@@ -160,14 +183,22 @@ protected void replaceRemoteEntityInCache(K key, long eventVersion) {
-            // We received event from remoteCache, so we won't update it back
-            replaced = cache.getAdvancedCache().withFlags(Flag.SKIP_CACHE_STORE, Flag.SKIP_CACHE_LOAD, Flag.IGNORE_RETURN_VALUES)
-                    .replace(key, localEntityWrapper, sessionWrapper);
+            KeycloakModelUtils.runJobInTransaction(sessionFactory, (session -> {
+
+                RealmModel realm = session.realms().getRealm(sessionWrapper.getEntity().getRealmId());
+                long lifespanMs = lifespanMsLoader.apply(realm, sessionWrapper.getEntity());
+                long maxIdleTimeMs = maxIdleTimeMsLoader.apply(realm, sessionWrapper.getEntity());
```
Could this be extracted out of the `do`-`while` loop? Session creation is costly and it is unlikely that `lifespanMs` and `maxIdleTimeMs` would change during the loop.

hmlnarik replied:
The `KeycloakModelUtils.runJobInTransaction` can be seen inside the `do`-`while` loop. Could it be extracted completely?

mposolda replied:
Replied to this in the other comment https://github.com/keycloak/keycloak/pull/7736#discussion_r569302965 .

hmlnarik reviewed (commented):
Thanks @mposolda! Looks very good to me, I have only a few nits to check, please see inline. I would approve, yet the failures in the cross-dc tests deserve some investigation

mposolda commented:
@hmlnarik Thanks for the review! I've addressed your concerns besides the cross-dc failures, which I will try to investigate more...

martin-kanis commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProvider.java:
```diff
@@ -463,144 +462,17 @@ protected void removeUserSessions(RealmModel realm, UserModel user, boolean offl
+    public void removeAllExpired() {
+        // Rely on expiration of cache entries provided by infinispan. Just expire entries from persister is needed
+        // TODO: Avoid iteration over all realms here (Details in the KEYCLOAK-16802)
+        session.realms().getRealmsStream().forEach(this::removeExpired);
```
Are you going to change this within the PR or it's final? KEYCLOAK-16802 says "it will be fixed together with KEYCLOAK-16755" which is this PR.

mposolda replied:
Replied to this in the other place.

mposolda replied:
I don't plan to change this within this PR, but in the follow-up PR. The follow-up PR will be more focus on the persister and DB tuning, so I didn't want to introduce this into this PR (which is more focused on the infinispan) to make the scope of the PR smaller.

hmlnarik commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remotestore/RemoteCacheSessionListener.java:
```diff
@@ -156,18 +183,35 @@ protected void replaceRemoteEntityInCache(K key, long eventVersion) {
 
             SessionEntityWrapper<V> sessionWrapper = remoteSession.mergeRemoteEntityWithLocalEntity(localEntityWrapper);
 
+            RealmModel realm = realmRef.get();
+            if (realm == null) {
+
+                KeycloakModelUtils.runJobInTransaction(sessionFactory, (session -> {
+
+                    realmRef.set(session.realms().getRealm(sessionWrapper.getEntity().getRealmId()));
+
+                }));
+
+                realm = realmRef.get();
+            }
+
+            long lifespanMs = lifespanMsLoader.apply(realm, sessionWrapper.getEntity());
+            long maxIdleTimeMs = maxIdleTimeMsLoader.apply(realm, sessionWrapper.getEntity());
```
Could determining `lifespanMs` and `maxIdleTimeMs` be returned back into the `runJobInTransaction` function? The current state may not work if e.g. the `realm` is only a partially cached object that does not have lifespans defined and delegates the lookup lazily only when queried; however this would fail since the lazy loading needs an active transaction and the transaction currently ends before this lookup.

Disclaimer: We don't have it in the current codebase but it is possible, even with the current store, and it is one of the discussed optimization for the new store.

mposolda replied:
@hmlnarik Thanks for clarifying this. I did not consider this for the new store.

So the points to consider:
- lifespanMsLoader and maxIdleTimeMsLoader needs to be called within the transaction due the reason you mentioned
- Obtaining of RealmModel can't be easily moved outside of `do-while` to the top. As the ID of the realm is retrieved from the session, which is loaded inside the `do` block. With my latest change, the load of the RealmModel will be done only once (due the realmRef), however it won't work due the previous point
- Computing of `lifespanMs` and `maxIdleTimeMs` depends on the realm, but also on the sessionEntity and it's "lastSessionRefresh" (for UserSEssion) or "timestamp" (ClientSession). And this timestamps can change. For example in first iteration of `do` the session may have timestamp 1000, but in the second iteration, the timestamp can be 2000 (as there was an update from concurrent transaction, which updated the timestamp and which then caused this conflict and the fact that second iteration of `do` was needed).

Regarding this, I see the options are:
1) Wrap the whole `do-while` into the `runJobInTransaction` block
2) Revert to the behaviour as was before this change.
It seems the (1) is better due the fact that (2) can mean multiple transactions as you pointed (even if this shouldn't happen too often as it is not big chance that same session is updated concurrently, but it is indeed possibility).

Any better idea?

hmlnarik replied:
> Computing of lifespanMs and maxIdleTimeMs depends on the realm, but also on the sessionEntity and it's "lastSessionRefresh" (for UserSEssion) or "timestamp" (ClientSession). And this timestamps can change. For example in first iteration of do the session may have timestamp 1000, but in the second iteration, the timestamp can be 2000 (as there was an update from concurrent transaction, which updated the timestamp and which then caused this conflict and the fact that second iteration of do was needed).

I see, and you're right. If they have changed, then we need to start a separate transaction, otherwise the single transaction could always see the stale values and never succeed. For that reason alone, I think we should follow the Option 2. Apologies for confusion.

mposolda replied:
@hmlnarik Thanks! I've updated the PR to use the option 2.

mposolda commented:
@hmlnarik @martin-kanis Thanks for the comments. I've fixed the cross-dc test failures and hopefully the stability of the cross-dc tests should be increased now.

There is an interesting issue, that for HotRod protocol of version lower than 3, the "lifespan" and "maxIdle" arguments are treated as unix timep instead of the time interval, in case of the argument is bigger than 30 days (which exactly is the default time for offline sessions in Keycloak). See the https://infinispan.org/docs/stable/titles/hotrod_protocol/hotrod_protocol.html and search for "30 days" for the details.

I've added util method InfinispanUtil.toHotrodTimeMs and some places, which calls "put", "putIfAbsent" or "replace" with arguments "lifespan" and "maxIdle" to use it.

hmlnarik reviewed (approved):
Thank you @mposolda, this is great addition!

mposolda commented:
@hmlnarik @martin-kanis Thanks for the review! 

I am attaching pipeline runs for the completeness:
https://keycloak-jenkins.com/view/pipelines/job/universal-test-pipeline-server/2570/
https://keycloak-jenkins.com/view/pipelines/job/universal-test-pipeline-adapters/2071/
It looks that no regressions there

Re-run of the universal-test-pipeline after the last change in RemoteCacheSessionListener from yesterday (just re-run cross-dc, cluster and base tests for this one as there was a change only in the RemoteCacheSessionListener, which affects only cross-dc):
https://keycloak-jenkins.com/view/pipelines/job/universal-test-pipeline-server/2581/

y-tabata commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/SessionTimeouts.java:
```diff
@@ -0,0 +1,293 @@
+        // clientSession max lifespan has preference if set
+        if (realm.getClientSessionMaxLifespan() > 0) {
+            sessionMaxLifespan = realm.getClientSessionMaxLifespan();
+        }
```
@mposolda Don't you need to consider these value (ClientSessionMaxLifespan, ClientSessionIdleTimeout, ClientOfflineSessionMaxLifespan, ClientOfflineSessionIdleTimeout) are overridden per client?

mposolda replied:
@y-tabata Thanks for pointing this. I've created https://issues.redhat.com/browse/KEYCLOAK-17111 for this. I see that before this PR, the InfinispanUserSessionProvider.removeExpired(realm) also ignored client-overriden timeouts, so this bug is not a regression of this PR. However it works fixing.
